render cms site wide notifications in react

umarmughal824 opened the pull request:
#### What are the relevant tickets?
https://trello.com/c/Du27kx7o/176-render-cms-site-wide-notifications-in-react

#### What's this PR do?
fixes #544 , Render cms site wide notifications in react

#### How should this be manually tested?
Just add notification from cms and then will be able to see the notifications on the site

#### Screenshots (if appropriate)
<img width="1439" alt="Screenshot 2020-06-05 at 00 17 43" src="https://user-images.githubusercontent.com/4043989/83801156-17b7a900-a6c2-11ea-8b73-39cec5b7bccf.png">
<img width="1440" alt="Screenshot 2020-06-03 at 17 47 26" src="https://user-images.githubusercontent.com/4043989/83642562-71d44380-a5c8-11ea-90a2-5b273954fb0c.png">
<img width="1440" alt="Screenshot 2020-06-03 at 18 32 26" src="https://user-images.githubusercontent.com/4043989/83642686-97f9e380-a5c8-11ea-9eb5-ac9931511166.png">

umarmughal824 commented:
@gsidebo I have addressed your feedback, could you please review it now.

gsidebo commented on static/js/components/NotificationContainer.js:
```diff
@@ -32,6 +32,15 @@ export class NotificationContainer extends React.Component<Props, State> {
     const { removeUserNotification, messageRemoveDelayMs } = this.props
     const { hiddenNotifications } = this.state
 
+    if (notificationKey === "cms-site-wide-notification") {
```
1. Please create a constant for this value (`"cms-site-wide-notification"`) and use it here and in App.js
1. Can you include this comment above the `if` statement?
    
    ```javascript
    // If the notification came from the CMS, set a local storage value to
    // avoid showing again.
    ```

gsidebo commented on static/js/pages/App.js:
```diff
@@ -19,50 +20,93 @@ import AccountSettingsPage from "./settings/AccountSettingsPage"
+export class App extends Component<Props> {
+  componentDidMount() {
+    const { addUserNotification } = this.props
+    const notifications = document.querySelector(".site-wide")
```
Nitpick: please rename this variable to `cmsNotification`

gsidebo commented on static/js/pages/App.js:
```diff
@@ -19,50 +20,93 @@ import AccountSettingsPage from "./settings/AccountSettingsPage"
+    const notifications = document.querySelector(".site-wide")
+    if (notifications) {
+      const notificationId = notifications.getAttribute("data-notification-id")
+      const notificationMessage = notifications.getElementsByTagName("p")[0]
```
Can we just use `notifications.innerText` or `notifications.textContent` here?

umarmughal824 replied:
that's is supposed to be the at the same line but linter is moving it to the next line.

umarmughal824 replied:
My build started to fail by moving it in the same line. It's asking to format it.

umarmughal824 replied:
Now going to revert it back.

umarmughal824 commented on static/js/pages/App.js:
```diff
@@ -19,50 +20,93 @@ import AccountSettingsPage from "./settings/AccountSettingsPage"
-    return <div className="app" />
+    return (
+      <Fragment>
+        {currentUser ? (
```
Earlier it was function-based components, now it has been converted to the class-based component so that type of rendering is not supported in class-based components.

umarmughal824 commented:
@gsidebo Regarding your comment on `static/js/notifications.js`, its not being rendered twice and cms notifictaion is working fine even without it.
 
But as its added in https://github.com/mitodl/bootcamp-ecommerce/pull/537, I don't know its other usage that's pretty big PR by @mbertrand . He can better answer this. Or we need a separate ticket to explore that.

mbertrand commented:
`static/js/notifications.js` was copied over from xpro along with `NotificationContainer.js` etc.  I assumed it was necessary for that component to work, but after some testing, that doesn't seem to be the case.  So it's probably safe to remove unless it's needed for the CMS notifications.

umarmughal824 commented:
thanks @mbertrand, for the clarification. It's not needed at all in cms notification. I was thinking so it might not be needed in the entire application because everything was seeming to be covered in the React component.

umarmughal824 commented:
@gsidebo I have done the following things.
- addressed your feedback
- removed notification.js
- added test cases.

Hope so, it should be good to go. Now waiting for your review.

gsidebo commented on static/js/components/NotificationContainer_test.js:
```diff
@@ -91,4 +91,50 @@ describe("NotificationContainer component", () => {
     })
   })
+
+  it("shows cms notifiction", async () => {
```
Nit:
```suggestion
  it("shows cms notification", async () => {
```

gsidebo commented on static/js/components/NotificationContainer_test.js:
```diff
@@ -91,4 +91,50 @@ describe("NotificationContainer component", () => {
+        userNotifications: {
+          "cms-site-wide-notification": {
+            type:  ALERT_TYPE_TEXT,
+            props: { text: "Cms Notifictaion", persistedId: 1 }
```
Nit:
```suggestion
            props: { text: "Cms Notification", persistedId: 1 }
```
This spelling occurs many times below

gsidebo commented on static/js/components/NotificationContainer_test.js:
```diff
@@ -91,4 +91,50 @@ describe("NotificationContainer component", () => {
+    const alerts = inner.find("Alert")
+    assert.lengthOf(alerts, 1)
+    const cmsNotifictaion = alerts.at(0)
+    assert.equal(cmsNotifictaion.prop("children").type, TextNotification)
```
We this `children` prop multiple times below. Let's store it in a variable called `notificationContents` or something like that
